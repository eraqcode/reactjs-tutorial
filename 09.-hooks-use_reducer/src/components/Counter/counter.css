.flex {
    display: flex;
    flex-direction: column;
    align-content: space-around;
    justify-items: center;
}

.btn-1 {
    margin-left: 5px;
    
}

.dark {
    font-weight: 700;
    font-size: 1.1rem;
}