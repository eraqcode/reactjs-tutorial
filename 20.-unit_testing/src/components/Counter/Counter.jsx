import React, { useState } from 'react'
import PropTypes from 'prop-types'
export const Counter = ( {value = 20} ) => {

    const [counter, setCounter] = useState( value )

    const handleIncrement = () => {
        setCounter( counter + 1 )
    }
    const handleReset = () => {
        setCounter( value )
    }
    const handleDecrement = () => {
        setCounter( counter - 1 )
    }
    return (
        <section>
            <header className="header-counter">
                <h2>My Counter</h2>
            </header>
            <div className="container">
                <h3> Count: { counter } </h3>
                <div className="buttons">
                    <button
                        className= "btn"
                        onClick={handleIncrement}
                    >
                        +
                    </button>
                    <button
                        className= "btn"
                        onClick={handleReset}
                    >
                        Reset
                    </button>
                    <button
                        className= "btn"
                        onClick={handleDecrement}
                    >
                        -
                    </button>
                </div>
            </div>
        </section>
    )
}

Counter.propTypes = {
    value: PropTypes.number
}