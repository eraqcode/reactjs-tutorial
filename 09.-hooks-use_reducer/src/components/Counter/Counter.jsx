import React from 'react'
import { useReducer } from 'react'


import './counter.css'
export const Counter = () => {

    const counter = 0; //initialState

    const counterReducer = (state, action) => {
        switch( action.type ) {
            case "increment":
                return state + 1;
            case "decrement":
                return state - 1;
            case "increment 5":
                return state + action.payload;
            case "decrement 5":
                return state - action.payload;        
            case "reset":
                return counter;
            default:
                return state;            
        }
    }

    const [ count, dispatch ] = useReducer( counterReducer, counter )

    return (
        <section>
            <header>
                <h3 className="blue-text">Counter Exercise</h3>
                <hr/>
            </header>
            <div className="row">
                <div className="col s6">
                    <h3> Counter: { count } </h3>
                </div>
                <div className="col s6 flex">
                    <div className="row flex">
                        <button
                            className=" waves-effect waves-orange waves-ripple btn blue darken-2"
                            onClick={ () => dispatch({ type: "increment" }) }
                        >
                            Increment
                        </button>
                        <button
                            className=" waves-effect waves-orange waves-ripple btn blue darken-2"
                            onClick={ () => dispatch({ type: "decrement" }) }
                        >
                            Decrement
                        </button>
                        <button
                            className=" waves-effect waves-orange waves-ripple btn blue darken-2"
                            onClick={ () => dispatch({ type: "reset" }) }
                        >
                            Reset
                        </button>
                    </div>

                    <div className="row flex">
                        <button
                            className="waves-effect waves-light btn yellow darken-2 black-text dark"
                            onClick={ () => dispatch({ type: "increment 5", payload: 5 }) }
                        >
                            Increment +5
                        </button>
                        <button
                            className="waves-effect waves-light btn yellow darken-2 black-text dark"
                            onClick={ () => dispatch({ type: "decrement 5", payload: 5 }) }
                        >
                            Decrement -5
                        </button>
                        <button
                            className="waves-effect waves-light btn yellow darken-2 black-text dark"
                            onClick={ () => dispatch({ type: "reset" }) }
                        >
                            Reset
                        </button>
                    </div>
                </div>
            </div>
        </section>
    )
}
