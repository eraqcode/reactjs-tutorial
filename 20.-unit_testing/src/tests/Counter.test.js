import React from 'react';
import { shallow } from 'enzyme';
import "@testing-library/jest-dom";
import { Counter } from '../components/Counter/Counter';

describe('Name of the group', () => {
    
    let wrapper = shallow(
        <Counter />
    )
    beforeEach( () => {
        wrapper = shallow(
           <Counter />
       )
    } )



    test('show the <Counter /> component correctly', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('should show by default a value equal to 100', () => {
        
        const wrapper = shallow(
            <Counter value={100} />
        )

        const counterText = wrapper.find('h3').text().trim();
        expect( counterText ).toBe(`Count: 100`)

    });

    //simulate event with buttons
    test('increment +1', () => {
        
        wrapper.find('button').at(0).simulate('click');
        const counterText = wrapper.find('h3').text().trim();

        expect(counterText).toBe(`Count: 21`)

    }); 

    test('decrement -1', () => {
        
        wrapper.find('button').at(2).simulate('click');
        const counterText = wrapper.find('h3').text().trim();

        expect(counterText).toBe(`Count: 19`)

    }); 

    test('reset to the default value', () => {
        
        
        wrapper.find('button').at(0).simulate('click');
        wrapper.find('button').at(0).simulate('click');

        //reference to the reset button
        wrapper.find('button').at(1).simulate('click');

        const counterText = wrapper.find('h3').text().trim();

        expect( counterText ).toBe(`Count: 20`)
        console.log(counterText)

    }); 

});