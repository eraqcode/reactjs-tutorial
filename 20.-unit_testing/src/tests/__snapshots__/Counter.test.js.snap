// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Name of the group show the <Counter /> component correctly 1`] = `
<section>
  <header
    className="header-counter"
  >
    <h2>
      My Counter
    </h2>
  </header>
  <div
    className="container"
  >
    <h3>
       Count: 
      20
       
    </h3>
    <div
      className="buttons"
    >
      <button
        className="btn"
        onClick={[Function]}
      >
        +
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Reset
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        -
      </button>
    </div>
  </div>
</section>
`;
